lib/src/Pulsar/AppState.hs

```diff
@@ -24,88 +27,73 @@ type Worker = (Async (), MVar ())
+  , _appWorkers :: [Worker]                           -- a list of workers for consumers and producers that run in the background
+  , _appResponse :: [(ReqId, MVar Response)]          -- a list of registered requests that need a Request Id
+  , _appSendReceipts :: [(ProducerId, ProducerSeqs)]  -- a list of registered messages sent by a specific producer
   }
+$(makeLenses ''AppState)
 
 mkConsumerId :: MonadIO m => Chan Response -> IORef AppState -> m ConsumerId
-mkConsumerId chan ref = liftIO $ atomicModifyIORef
-  ref
-  (\(AppState cs cid pid rid w rs s) ->
-    let cid' = cid + 1
-    in  (AppState ((cid, chan) : cs) cid' pid rid w rs s, cid)
-  )
+mkConsumerId chan ref = liftIO $ atomicModifyIORef ref $ \app ->
+  let cid = app ^. appConsumerId
+      f   = over appConsumers ((cid, chan) :) app
+  in  (over appConsumerId (+ 1) f, cid)
 
 mkProducerId :: MonadIO m => IORef AppState -> m ProducerId
-mkProducerId ref = liftIO $ atomicModifyIORef
-  ref
-  (\(AppState cs cid pid rid w rs s) ->
-    let pid' = pid + 1 in (AppState cs cid pid' rid w rs ((pid, []) : s), pid)
-  )
-
-registerSeqId
-  :: MonadIO m => IORef AppState -> ProducerId -> SeqId -> m (MVar Response)
-registerSeqId ref pid sid = liftIO $ do
-  var <- newEmptyMVar
-  atomicModifyIORef
-    ref
-    (\(AppState cs cid pd req w rs xs) ->
-      let foo = filter (\(p, _) -> pid == p) xs >>= snd
-          bar = (sid, var) : foo
-          taz = (\(p, ys) -> if p == pid then (p, bar) else (p, ys)) <$> xs
-      in  (AppState cs cid pd req w rs taz, var)
-    )
+mkProducerId ref = liftIO $ atomicModifyIORef ref $ \app ->
+  let pid = app ^. appProducerId
+      f   = over appSendReceipts ((pid, []) :) app
+  in  (over appProducerId (+ 1) f, pid)
 
 mkRequestId :: MonadIO m => IORef AppState -> m (ReqId, MVar Response)
 mkRequestId ref = liftIO $ do
   var <- newEmptyMVar
-  atomicModifyIORef
-    ref
-    (\(AppState cs cid pid req w rs s) ->
-      let req' = req + 1
-      in  (AppState cs cid pid req' w ((req, var) : rs) s, (req, var))
-    )
+  atomicModifyIORef ref $ \app ->
+    let req = app ^. appRequestId
+        f   = over appResponse ((req, var) :) app
+    in  (over appRequestId (+ 1) f, (req, var))
 
 addWorker :: MonadIO m => IORef AppState -> (Async (), MVar ()) -> m ()
-addWorker ref nw = liftIO $ atomicModifyIORef
-  ref
-  (\(AppState cs cid pid req w rs s) ->
-    (AppState cs cid pid req (nw : w) rs s, ())
-  )
+addWorker ref nw =
+  liftIO $ atomicModifyIORef ref $ \app -> (over appWorkers (nw :) app, ())
 
 {- | Register a response for a request and unregister request. -}
 registerReqResponse :: MonadIO m => IORef AppState -> ReqId -> Response -> m ()
 registerReqResponse ref rid resp = liftIO $ do
-  maybeVar <- atomicModifyIORef
-    ref
-    (\(AppState cs cid pid req w rs s) ->
-      (AppState cs cid pid req w (h rs) s, g rs)
-    )
+  maybeVar <- atomicModifyIORef ref
+    $ \app -> (over appResponse h app, lookup rid $ app ^. appResponse)
```

Comment:
Is there a `lookupOf` combinator in the lens library?

`Getter s (k, v) -> k -> s -> Maybe v`

Reply:
I couldn't find anything with that type signature, this lens library is weird af :smile: See http://hackage.haskell.org/package/lens-family-core-2.1.0/docs/Lens-Family.html & http://hackage.haskell.org/package/lens-family-core-2.1.0/docs/Lens-Family-Stock.html

But I don't want to use `lens` because `lens-family` is already in scope due to the `proto-lens` dep and that would create a bigger footprint on potential users... Though, I wish they have used `lens` instead...